### DISCUSSION TITLE
Building a macOS sysroot from bottles (in Ubuntu containers)

### DISCUSSION THREAD

**Discussion by jphastings** [Homebrew/discussions#3784] [Everyday usage] [answered]
Quick context: [GoReleaser](https://goreleaser.com/) lets you [cross-compile Go/CGo applications](https://goreleaser.com/cookbooks/cgo-and-crosscompiling/) for a wide variety of platforms using a [cross-compiling Docker container](https://github.com/goreleaser/goreleaser-cross). That container supports custom "sysroots", containing the necessary libraries for linking. ([Example repo](https://github.com/goreleaser/goreleaser-cross-example#sysroot) using a [Homebrew-based](https://github.com/goreleaser/goreleaser-cross-example-sysroot/tree/master/macos/amd64/usr/local) sysroot; and my own [homebrew sysroot](https://github.com/dotpostcard/postcards-go/tree/main/sysroots/macos) for this tool)

---

I recognise that if I want to _build_ homebrew formula I'm pretty much certainly going to need an actual macOS machine, but would it be a smart move for me to build a `brewfile2sysroot` tool that would allow me to prepare a sysroot with the libraries needed to build a CGo app, but runnable on linux by downloading _bottles_?

I imagine the steps would be:
1. Accept a list of homebrew dependencies from a `Brewfile`
2. Download the bottles for those formula
3. Unpack those bottles to an appropriate location
4. Set whatever env vars are appropriate for [goreleaser-cross](https://github.com/goreleaser/goreleaser-cross) to make use of them
5. (And be able to do the above for both amd64 and arm64 bottles, irrespective of the host architecture)

Should I:
- Parse the Bottle URLs out of the Formulae in the git repo, and extract them manually (fragile, because I'd be using a private API?), or
- Install `homebrew` on my linux box, but trick it into thinking its running on macOS, so it installs macOS bottles, not the linux ones (fragile, as some formulae might require unusual interactivity with the host system)
- Something else, because I've not thought this through enough?

(As an aside, I think I could also _build_ formula on linux by using [sickcodes/Docker-OSX](https://github.com/sickcodes/Docker-OSX) — which looks seriously interesting — but a 20GB container download is intense for a build!)

**Comment by Bo98** [comment, marked as answer]
The short answer is there's not an easy way right now. Or certainly not something that is fully supported.

> Install homebrew on my linux box, but trick it into thinking its running on macOS

We do allow a small level of simulation (`HOMEBREW_SIMULATE_MACOS_ON_LINUX`). However this env exists for parsing dep trees or other basic metadata about a formula rather than actually installing.

Ignoring the `OS.mac?`/`OS.linux?` checks (based on `uname`) that will make it install Linux bottles, the bottle install code actually does a fairly good job at not calling much system tools so could technically work on Linux. Maybe some pre-install system environment checks could trip you up.

The biggest issue would be post-install. That's a free-for-all that individual formulae can do whatever they want. For example, `ca-certificates` will use the system's `security` command-line tool to extract system CA certificates and add it to the just-installed CA bundle. Not every formula has a post-install, it's only a few that do.

> Parse the Bottle URLs out of the Formulae in the git repo, and extract them manually (fragile, because I'd be using a private API?),

There are one of the API endpoints:

* https://formulae.brew.sh/api/formula.json (full formula information on everything)
* https://formulae.brew.sh/api/formula/hello.json (full formula information on a single formula)
  - Also available as `brew info --json=v2 --formula hello`
* https://formulae.brew.sh/api/bottle/hello.json (just the bottle information of a single formula)
  - Also avilable as `brew info --json=v2 --bottle --formula hello`

Note that GitHub Packages require a `Bearer` Authorization header, but the value can be totally bogus.

Something also available is `brew fetch --bottle_tag=monterey hello`. Though where it saves to is not particularly publicly documented. It's printed in the output, though as message designed for humans so the output isn't guaranteed to remain constant.

While the contents of the tarball I guess is technically not documented, in reality it's never going to change. It directly matches the install layout on disk, with the `etc` and `var` being stored in `.bottle` which is moved out of Cellar and has special logic to not overwrite the user's configs.

The problem though is included binaries and scripts are not executable by default. Run `otool -L` on one of the binaries and you will see they contain references to `@@HOMEBREW_PREFIX@@` rather than real path. Homebrew would usually, when installing, replace these for you. The reason for these existing is some bottles support being installed to any prefix rather than just the default, and this is the mechanism that helps make that possible.

Exactly what gets replaced can increase over time. For example, for script files we added `@@HOMEBREW_PERL@@` and `@@HOMEBREW_JAVA@@` somewhat recently to modify the shebangs of Perl and OpenJDK dependents based on the OS version. This list is not exhaustive - there are a couple more, mostly for text-like files.

In the future we may do more things like binary patching to achieve full relocatability, which is available as a developer test.

**Reply by jphastings** [reply]
This is a superb answer, thank you! It seems like there's probably a hacky way forward here, but nothing that's likely to be as seemless as `brew`, which is certainly something I'd want to aspire to.

Though I'm short on time at the moment, I'm tempted to go down the route of building a `homebrew-sysroot` command, as I imagine I'd have slightly more access to the internal APIs that you mention above. At the least, I should be able to be more explicit about the features unsupported ("warning `ca-certificates` is attempting to run a post-install script, which isn't supported on this platform" kinda thing).

You've given me plenty to think about, I very much appreciate your time!

**Reply by carlocab** [reply]
You might wish to consider populating your sysroot with text-base `.tbd` stubs instead of the libraries themselves. I believe this should provide the information you need for linking when cross-compiling.

Of course, you'll still need to construct those `.tbd` stubs somehow, but it might be enough to use a standard tool (maybe [this one](https://github.com/tpoechtrager/apple-libtapi)?) on the downloaded bottles, and then fix up the `@HOMEBREW_PREFIX@` references after.

The benefit of this approach is that you don't really need to reach too far into `brew` internals to convince it to install a macOS bottle on Linux, plus the resulting assembled sysroot will take up much less storage.

You'll need a linker that can read `.tbd` stubs, but you've probably already got one of those.

Note: I've never done this, so I don't actually know for certain that it'll work.

**Reply by Bo98** [reply]
`.tbd` is quite a good idea. You could probably generate one on a binary directly from the bottle without needing to deal with the relocation.
